src/storage/restrict.rs

```diff
@@ -0,0 +1,249 @@
+        entry.assert_same_storage(self.data.borrow());
+        unsafe { self.data.borrow_mut().get_mut(entry.index()) }
+    }
+}
+
+impl<'rf, 'st, B, T, R> RestrictedStorage<'rf, 'st, B, T, R, restrict_type::Normal>
+    where T: Component,
+          R: BorrowMut<T::Storage>,
+          B: Borrow<BitSet>,
+{
+    /// Attempts to get the component related to the entity mutably.
+    ///
+    /// Functions similar to the normal `Storage::get_mut` implementation.
+    pub fn get_mut(&mut self, entity: Entity) -> Option<&mut T> {
+        if self.bitset.borrow().contains(entity.id()) && self.entities.is_alive(entity) {
+            Some(unsafe { self.data.borrow_mut().get_mut(entity.id()) })
+        } else {
+            None
+        }
+    }
+
+}
+
+impl<'rf, 'st: 'rf, B, T, R, RT> Join for &'rf RestrictedStorage<'rf, 'st, B, T, R, RT>
+    where T: Component,
+          R: Borrow<T::Storage>,
+          B: Borrow<BitSet>,
+{
+    type Type = (Entry<'rf, T>, Self); 
+    type Value = Self;
+    type Mask = &'rf BitSet;
+    fn open(self) -> (Self::Mask, Self::Value) {
+        (self.bitset.borrow(), self)
+    }
+    unsafe fn get(value: &mut Self::Value, id: Index) -> Self::Type {
+        let entry = Entry {
+            id: id,
+            pointer: value.data.borrow() as *const T::Storage,
+            phantom: PhantomData,
+        };
+        
+        (entry, value) // reference?
+    }
+}
+
+impl<'rf, 'st: 'rf, B, T, R, RT> Join for &'rf mut RestrictedStorage<'rf, 'st, B, T, R, RT>
+    where T: Component,
+          R: BorrowMut<T::Storage>,
+          B: Borrow<BitSet>,
+{
+    type Type = (Entry<'rf, T>, Self); 
+    type Value = Self;
+    type Mask = BitSet;
+    fn open(self) -> (Self::Mask, Self::Value) {
+        (self.bitset.borrow().clone(), self)
+    }
+    unsafe fn get(value: &mut Self::Value, id: Index) -> Self::Type {
+        use std::mem;
+        let entry = Entry {
+            id: id,
+            pointer: value.data.borrow() as *const T::Storage,
+            phantom: PhantomData,
+        };
+
+        let value: &'rf mut Self::Value = mem::transmute(value);
+        (entry, value)
+    }
+}
+
+impl<'st, T, D> Storage<'st, T, D>
+    where T: Component,
+          D: Deref<Target = MaskedStorage<T>>,
+{
+    /// Builds an immutable `RestrictedStorage` out of a `Storage`. Allows restricted
+    /// access to the inner components without allowing invalidating the
+    /// bitset for iteration in `Join`.
+    pub fn restrict<'rf>(&'rf self)
+        -> RestrictedStorage<'rf, 'st, &'rf BitSet, T, &'rf T::Storage, restrict_type::Normal>
+    {
+        RestrictedStorage {
+            bitset: &self.data.mask,
+            data: &self.data.inner,
+            entities: &self.entities,
+            phantom: PhantomData,
+        }
+    }
+
+    /// Builds a parallel `RestrictedStorage`, does not allow mutably getting other components
+    /// aside from the current iteration.
+    pub fn par_restrict<'rf>(&'rf self)
```

Comment:
Do we really need to methods for immutable restriction storages? Because they both don't allow mutable access, I don't see how their functionality is different.